pkg/kstatus/observe/observer/observer.go

```diff
@@ -0,0 +1,197 @@
+package observer
+
+import (
+	"context"
+	"time"
+
+	"k8s.io/apimachinery/pkg/api/meta"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"sigs.k8s.io/cli-utils/pkg/kstatus/observe/event"
+	"sigs.k8s.io/cli-utils/pkg/kstatus/wait"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// AggregatorFactoryFunc defines the signature for the function the Observer will use to
+// create a new StatusAggregator for each StatusObserverRunner.
+type AggregatorFactoryFunc func(identifiers []wait.ResourceIdentifier) StatusAggregator
+
+// ReaderFactoryFunc defines the signature for the function the Observer will use to create
+// a new ObserverReader for each StatusObserverRunner.
+type ReaderFactoryFunc func(reader client.Reader, mapper meta.RESTMapper, identifiers []wait.ResourceIdentifier) (ObserverReader, error)
+
+// ObserversFactoryFunc defines the signature for the function the Observer will use to
+// create the resource observers and the default observer for each StatusObserverRunner.
+type ObserversFactoryFunc func(reader ObserverReader, mapper meta.RESTMapper) (resourceObservers map[schema.GroupKind]ResourceObserver, defaultObserver ResourceObserver)
+
+// Observer provides functionality for polling a cluster for status for a set of resources.
+type Observer struct {
+	Reader client.Reader
```

Comment:
Are their sane defaults we can use for these if they aren't specified by the user?

Reply:
This is the interface to the core framework. I will add a separate interface that is meant for end-users that will provide defaults for many of these fields.